tests/shared/src/test/scala/cats/tests/HashMapSuite.scala

```diff
@@ -0,0 +1,292 @@
+    }
+  }
+
+  property("Empty HashMap add consistent with contains") {
+    forAll { (i: Int, s: String) =>
+      assert(HashMap.empty[Int, String].add(i, s).contains(i))
+    }
+  }
+
+  property("Empty HashMap add and remove consistent with contains") {
+    forAll { (i: Int, s: String) =>
+      assert(!HashMap.empty[Int, String].add(i, s).remove(i).contains(i))
+    }
+  }
+
+  property("fromSeq consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+
+      kvs.foreach { case (k, _) =>
+        assert(hashMap.contains(k))
+      }
+    }
+  }
+
+  property("remove consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val distinctKvs = distinctBindings(kvs)
+      val hashMap = HashMap.fromSeq(kvs)
+
+      distinctKvs.foldLeft(hashMap) { case (hm, (i, _)) =>
+        if (!hm.contains(i)) {
+          println(hm)
+          println(i)
+          println(hm.contains(i))
+        }
+        assert(hm.contains(i))
+        assert(!hm.remove(i).contains(i))
+        hm.remove(i)
+      }
+
+      ()
+    }
+  }
+
+  property("add and remove with collisions consistent with contains") {
+    forAll(collidingKvs) { (collisions: List[((String, Int), (String, Int))]) =>
+
+      val distinctCollisions = collisions.distinctBy(_._1._1)
+
+      val hashMap = distinctCollisions.foldLeft(HashMap.empty[String, Int]) { case (hm, ((lk, lv), (rk, rv))) =>
+        hm.add(lk, lv).add(rk, rv)
+      }
+
+      distinctCollisions.foldLeft(hashMap) { case (hm, ((lk, _), (rk, _))) =>
+        assert(hm.contains(lk))
+        assert(hm.contains(rk))
+
+        val removeL = hm.remove(lk)
+        assert(removeL.contains(rk))
+        assert(!removeL.contains(lk))
+
+        val removeR = hm.remove(rk)
+        assert(removeR.contains(lk))
+        assert(!removeR.contains(rk))
+
+        val removeLR = removeL.remove(rk)
+        assert(!removeLR.contains(lk))
+        assert(!removeLR.contains(rk))
+
+        removeLR
+      }
+
+      ()
+    }
+  }
+
+  property("iterator consistent with contains") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+      val distinctKvs = distinctBindings(kvs)
+
+      val iterated = mutable.ListBuffer[(Int, String)]()
+      hashMap.iterator.foreach { iterated += _ }
+
+      assert(distinctKvs.forall(iterated.contains))
+      assert(iterated.forall(distinctKvs.contains))
+      assert(iterated.toList.distinctBy(_._1) === iterated.toList)
+      assert(iterated.forall { case (k, _) => hashMap.contains(k) })
+    }
+  }
+
+  property("iterator consistent with reverseIterator") {
+    forAll { (kvs: List[(Int, String)]) =>
+      val hashMap = HashMap.fromSeq(kvs)
+
+      val iterated = mutable.ListBuffer[(Int, String)]()
+      val reverseIterated = mutable.ListBuffer[(Int, String)]()
+      hashMap.iterator.foreach { iterated += _ }
+      hashMap.reverseIterator.foreach { reverseIterated += _ }
```

Comment:
I wonder why we have `reverseIterator` when the order is not a contract we can give. I wonder if we should just remove this.

Reply:
In the standard library the `reverseIterator` is used to implement operations like `init` and `last` that don't naturally make a lot of sense for a HashMap, so there may not be any point in keeping it in cats.